nemo/collections/nlp/models/nlp_model.py

```diff
@@ -312,53 +252,225 @@ def setup(self, stage: str) -> None:
+                        f'and checkpoint version: {checkpoint_version}'
+                    )
+                )
+            os.chdir(cwd)
+
+        if app_state.model_parallel_size is not None:
+            if checkpoint_version is None:
+                raise ValueError(
+                    "Restoring from megatron model parallel .nemo but could not find megatron checkpoint version."
+                )
+            else:
+                logging.info(f"Setting megatron checkpoint version: {checkpoint_version}")
+                set_checkpoint_version(checkpoint_version)
+
+            app_state.world_size = trainer.num_gpus * trainer.num_nodes
+
+            # try to get local rank from global
+            local_rank = None
+            try:
+                local_rank = int(os.environ['LOCAL_RANK'])
+            except:
+                logging.info('Global variable LOCAL_RANK not yet specified. Assuming LOCAL_RANK is 0.')
+
+            if local_rank is not None:
+                app_state.local_rank = local_rank
+            else:
+                # if local is None then we are on the main process
+                local_rank = 0
+
+            model_parallel_rank = compute_model_parallel_rank(local_rank, app_state.model_parallel_size)
+            app_state.model_parallel_rank = model_parallel_rank
+
+            restored_model = cls._default_restore_from(
+                restore_path, override_config_path, map_location, strict, return_config
+            )
+            restored_model._trainer = trainer
+            return restored_model
+        else:
+            return super().restore_from(restore_path, override_config_path, map_location, strict, return_config)
+
+    @rank_zero_only
+    def register_megatron_checkpoint_version(self):
+        """ Adds checkpoint version to .nemo archive """
+        if self.bert_model is None:
+            raise ValueError('Instantiate self.bert_model before registering megatron checkpoint version.')
+        else:
+            # get encoder config and create source for artifact
+            if isinstance(self.bert_model, MegatronBertEncoder):
+                checkpoint_version = get_checkpoint_version()
+                if checkpoint_version is None:
+                    raise ValueError('Unable to get megatron checkpoint version.')
+                else:
+                    checkpoint_version_dict = {'checkpoint_version': checkpoint_version}
+                    checkpoint_version_path = 'megatron_checkpoint_version.json'
+                    checkpoint_version_src = os.path.join(NEMO_NLP_TMP, checkpoint_version_path)
+                    with open(checkpoint_version_src, 'w') as f:
+                        f.write(json.dumps(checkpoint_version_dict))
+                    self.register_artifact(checkpoint_version_path, checkpoint_version_src)
+
+    @staticmethod
+    def __unpack_nemo_file(path2file: str, out_folder: str) -> str:
```

Comment:
isnt this same as ModelPT impl? Rather replace the `__unpack_nemo_file` with `_unpack_nemo_file` and call super?

Reply:
I tried so many ways but could not make it work. Because it's a class method and a double underscore, nothing I did would work.

Reply:
Oh I meant change the ModelPT double underscore to single underscore, and use it here. Theres no particular reason the ModelPT impl of this should be name mangled and unavailable to its subclasses.